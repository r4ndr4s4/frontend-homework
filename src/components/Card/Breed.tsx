import Card from '@mui/material/Card';
import CardHeader from '@mui/material/CardHeader';
import CardMedia from '@mui/material/CardMedia';
import CardContent from '@mui/material/CardContent';
import CardActions from '@mui/material/CardActions';
import IconButton from '@mui/material/IconButton';
import Typography from '@mui/material/Typography';
import FavoriteIcon from '@mui/icons-material/Favorite';

import NotFound from '../NotFound';
import Container from '../Container';
import useBreed from '@/hooks/useBreed';
import { useAppDispatch, useAppSelector } from '@/app/store';
import { add, remove } from '@/features/favoritesSlice';

function Breed({ breedId }: { breedId: number }) {
  const { isPending, error, breed } = useBreed(breedId);

  const dispatch = useAppDispatch();

  const isFavorite = useAppSelector(({ favorites }) => favorites.breedIds.includes(breedId));

  if (isPending) {
    return 'Loading...';
  }

  // TODO add error boundary
  if (error) {
    return 'An error has occurred: ' + error.message;
  }

  if (!breed) {
    return <NotFound entity="Breed" />;
  }

  return (
    <Container>
      <Card>
        <CardHeader title={breed.name} subheader={breed.breed_group} />
        <CardMedia
          component="img"
          image={`https://cdn2.thedogapi.com/images/${breed.reference_image_id}.jpg`}
          alt={breed.name}
        />

        <CardContent>
          <Typography variant="body2" sx={{ color: 'text.secondary' }}>
            {breed.bred_for}. {breed.life_span}. {breed.weight.metric} kgs, {breed.height.metric} cms.{' '}
            {breed.temperament}.
          </Typography>
        </CardContent>

        <CardActions>
          <IconButton
            aria-label="add to favorites"
            onClick={() => dispatch(isFavorite ? remove(breedId) : add(breedId))}
          >
            <FavoriteIcon sx={{ color: isFavorite ? 'Red.Base' : 'inherit' }} />
          </IconButton>
        </CardActions>
      </Card>
    </Container>
  );
}

export default Breed;
