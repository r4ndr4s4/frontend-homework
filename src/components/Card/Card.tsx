import MaterialCard from '@mui/material/Card';
import CardHeader from '@mui/material/CardHeader';
import CardMedia from '@mui/material/CardMedia';
import CardContent from '@mui/material/CardContent';
import CardActions from '@mui/material/CardActions';
import IconButton from '@mui/material/IconButton';
import Typography from '@mui/material/Typography';
import FavoriteIcon from '@mui/icons-material/Favorite';

import { Breed as BreedType } from '@/utils/types';

function Card({
  breed,
  isFavorite,
  handleFavoriteChange,
}: {
  breed: BreedType;
  isFavorite: boolean;
  handleFavoriteChange: () => void;
}) {
  return (
    <MaterialCard>
      <CardHeader title={breed.name} subheader={breed.breed_group} />
      <CardMedia
        component="img"
        image={`https://cdn2.thedogapi.com/images/${breed.reference_image_id}.jpg`}
        alt={breed.name}
      />

      <CardContent>
        <Typography variant="body2" sx={{ color: 'text.secondary' }}>
          {breed.bred_for}. {breed.life_span}. {breed.weight.metric} kgs, {breed.height.metric} cms. {breed.temperament}
          .
        </Typography>
      </CardContent>

      <CardActions>
        <IconButton aria-label="add to favorites" onClick={handleFavoriteChange}>
          <FavoriteIcon sx={{ color: isFavorite ? 'Red.Base' : 'inherit' }} />
        </IconButton>
      </CardActions>
    </MaterialCard>
  );
}

export default Card;
