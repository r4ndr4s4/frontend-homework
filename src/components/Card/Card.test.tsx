import { describe, expect, test } from 'vitest';
import { render, screen } from '@testing-library/react';

import Card from './Card';

const breed = {
  weight: {
    imperial: '40 - 65',
    metric: '18 - 29',
  },
  height: {
    imperial: '21 - 23',
    metric: '53 - 58',
  },
  id: 4,
  name: 'Airedale Terrier',
  bred_for: 'Badger, otter hunting',
  breed_group: 'Terrier',
  life_span: '10 - 13 years',
  temperament: 'Outgoing, Friendly, Alert, Confident, Intelligent, Courageous',
  origin: 'United Kingdom, England',
  reference_image_id: '1-7cgoZSh',
};

describe('Card', () => {
  test('renders', () => {
    const { container: result } = render(<Card breed={breed} isFavorite handleFavoriteChange={() => true} />);

    expect(screen.getByText(breed.name)).toBeDefined();

    expect(result).toMatchSnapshot();
  });
});
