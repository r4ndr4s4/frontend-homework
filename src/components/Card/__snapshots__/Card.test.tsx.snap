// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card > renders 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-nv68gf-MuiPaper-root-MuiCard-root"
    style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
  >
    <div
      class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-content css-1dpseql-MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-ku7d13-MuiTypography-root"
        >
          Airedale Terrier
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiCardHeader-subheader css-s619mj-MuiTypography-root"
        >
          Terrier
        </span>
      </div>
      <div
        class="MuiCardHeader-action css-sgoict-MuiCardHeader-action"
      >
        <button
          aria-label="add to favorites"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1wjzr1s-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-gr4rxj-MuiSvgIcon-root"
            data-testid="FavoriteIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54z"
            />
          </svg>
        </button>
      </div>
    </div>
    <img
      alt="Airedale Terrier"
      class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-squjdq-MuiCardMedia-root"
      src="https://cdn2.thedogapi.com/images/1-7cgoZSh.jpg"
    />
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body2 css-atovpb-MuiTypography-root"
      >
        Breed for: 
        <strong>
          Badger, otter hunting
        </strong>
        .
      </p>
      <p
        class="MuiTypography-root MuiTypography-body2 css-atovpb-MuiTypography-root"
      >
        Life expectancy: 
        <strong>
          10 - 13 years
        </strong>
        .
      </p>
      <p
        class="MuiTypography-root MuiTypography-body2 css-atovpb-MuiTypography-root"
      >
        Average weight: 
        <strong>
          18 - 29
           kgs
        </strong>
        .
      </p>
      <p
        class="MuiTypography-root MuiTypography-body2 css-atovpb-MuiTypography-root"
      >
        Average height: 
        <strong>
          53 - 58
           cms
        </strong>
        .
      </p>
      <p
        class="MuiTypography-root MuiTypography-body2 css-atovpb-MuiTypography-root"
      >
        Temperament: 
        <strong>
          Outgoing, Friendly, Alert, Confident, Intelligent, Courageous
        </strong>
        .
      </p>
    </div>
  </div>
</div>
`;
